### ISSUE TITLE
Proposal: Making the cmake build distribution friendly

### ISSUE THREAD

**Issue by dmacvicar** [tensorflow/tensorflow#13061]
The following is a proposal, and I don't have it fully working yet, so a Pull Request is too early, and I want to gather some feedback. If an issue is not appropriate, let me know and I will look for a different medium.

### System information

Almost any Linux distribution.

### Describe the problem

Most Linux distributions have similar policies/limitations:

1. Not accepting bundling of system libraries (specially those security-sensitive).
2. Requiring building the whole from source and allowing patching the sources.
3. Requiring building offline (without Internet access).
4. Not having enough manpower to package hundred of packages in a dependency chain just to get packages like maven built in these source-chains environments.

Because of 4., cmake makes things easier. It is a build tool with a few dependencies that does not need complex bootstrapping. Building bazel itself will bring you into the Java maven dependency chain which we already proved in openSUSE to expand into hundred of packages.

Both the bazel and cmake build files do not play well with 1. 2. 3., but cmake already improves 4.

The cmake build files, also pull the sources of different projects. That is because the top `CMakeLists.txt` includes:

```cmake
set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/external)
```

into the load path.

that external/ directory is full of snippets like:

* `tensorflow/contrib/cmake/external/jpeg.cmake`
* `tensorflow/contrib/cmake/external/gif.cmake`
* ...
* etc

Those snippets use the [external project](https://cmake.org/cmake/help/v3.0/module/ExternalProject.html) cmake api to build directly from git or tarballs upstream. This clashes with
1., 2., 3.

It is a bit sad that things like curl used to be looked up in the system and developers deliberately bundled them without making the cmake snippet first look for it, and if not, configure the bundled one.

### Proposal

The proposal is to make the cmake build support both the Google/Mac user type of build with bundled sources, and the classical Linux distribution build.
This would be a gradual refactoring. Steps could be:

* Add to CMakeLists.txt an option:

```cmake
option(tensorflow_SYSTEM_LIBRARIES "Use system libraries" ON)
```

* Replicate the external/ directory as platform/ and conditionally make it use one or the other:

```cmake
if(tensorflow_SYSTEM_LIBRARIES)
 set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/platform)
else()
  set(CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/external)
endif()
```

So that then the part that does:

```cmake
# External dependencies
include(zlib)
include(gif)
include(png)
include(jpeg)
```

would just work...

* Replace incrementaly and one by one the $dep.cmake snippets in `platform/`.

Example with gif.cmake:

Using the standard `/usr/share/cmake/Modules/FindGIF.cmake` included in cmake, which may use `pkg-config` for some modules. The module itself documents it would then set: `GIF_LIBRARIES`, `GIF_INCLUDE_DIR`, etc.

Unfortunately the variable that the original set is called `gif_STATIC_LIBRARIES`, because it assumes it would be static. I think we could fix that later so that naming ends making sense, but lets not focus on that for now.

The `gif.cmake` I cited above would become a simple:

```cmake
find_package(GIF REQUIRED)
# dummy targets other targets depend on.
add_custom_target(gif)
# These can be removed if we fix CMakeLists
add_custom_target(gif_copy_headers_to_destination)

set(gif_STATIC_LIBRARIES ${GIF_LIBRARIES})
set(gif_INCLUDE_DIR ${GIF_INCLUDE_DIR})
```

And that is more or less enough to move forward with this dependency... repeat.
It may need some hacks also in the top level cmake files however...

I think this approach is doable, and could be turned into making the cmake build fully Linux distro enabled upstream. Slowly cleaning the naming, etc, removing these copy_headers targets so that the platform/ version of the .cmake files do not need to create dummy targets, etc.

- Similarly, the sqlite one becomes:

```cmake
find_package(SQLite3)
set(sqlite_STATIC_LIBRARIES ${SQLITE3_LIBRARIES})
set(sqlite_HEADERS ${SQLITE3_INCLUDE_DIR})
add_custom_target(sqlite)
add_custom_target(sqlite_copy_headers_to_destination)
```

With the only difference is that the `Find` module was not included in cmake, so I just copied it into
platform from https://raw.githubusercontent.com/LuaDist/libsqlite3/master/cmake/FindSQLite3.cmake somewhere.

* With this method I have been able to move forward and forward with the build.

### Current blockers

* cmake build does not work out of the box [PR](https://github.com/tensorflow/tensorflow/pull/12734)
* [Issue](https://github.com/tensorflow/tensorflow/issues/12018) with `Missing tensorflow/core/debug/debug_service.grpc.pb.h`

**Comment by CDLuminate** [comment]
I second this proposal.

**Comment by mrry** [comment]
This is an area where we would warmly welcome contributions from the community. At present our only critical use for the `tensorflow/contrib/cmake` build is generating release binaries for TensorFlow on Windows. Anything that improves the general usefulness of that build, while preserving the ability to build our Windows release, would be most welcome.

**Comment by gunan** [comment]
Also, We are very close to moving windows to bazel, so we will lower the amount of support we will have available for cmake once that happens.
Therefore, I recommend using bazel instead for distributing TF.

**Comment by dmacvicar** [comment]
@gunan as I mentioned, given point 4.¹ bazel I see hard to to build with the requirements 1. 2. and 3. that most distributions have (maven is already a pain), so even if the bazel build files get improved to look for system libraries first, the problems with bazel, its bootstrapping and dragging of the maven ecosystem will prevail.

Do you see this different, that bazel can get properly packaged by Linux distributions? am I missing something here?.

**Comment by CDLuminate** [comment]
Bazel packaging for Debian seems stuck for a long time because bazel doesn't meet Debian's policy. One of the developers working on it has told me that it's hard to get it done.

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=782654

I think the only linux distribution that ships bazel, tensorflow (and cudnn) in its official archive is Archlinux. https://www.archlinux.org/packages/?sort=&q=tensorflow&maintainer=&flagged=

**Comment by dmacvicar** [comment]
@CDLuminate Archilinux has it because they [are not building it, they are packaging it](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/bazel), which means they are not applying any distribution policies to the build process (build from source, ability to patch, legal review of the sources, etc). Which is *exactly* my point. The package you install may have been built on someone else's workstation.

Every other distro will fall in the situation that as soon as you depend on a dependency built by maven, the dependency chain expands exponentially to hundred of packages, sometimes in multiple versions, and a very complicated bootstrap process. Bazel depends on itself _and_ on stuff build by maven.

cmake can be built with the base system (which is what I would expect from a build tool) and plays well with distributions with strict policies (openSUSE, Debian, etc).

**Comment by CDLuminate** [comment]
Let me add a citation to the proposal: https://www.debian.org/doc/debian-policy/index.html#main-building-script-debian-rules

```
For packages in the main archive, no required targets may attempt network access.
```

Note, the `required targets` include the whole package building process and more.

**Comment by yongtang** [comment]
I created a PR #13145 that should address the `Missing tensorflow/core/debug/debug_service.grpc.pb.h` issue on Linux for cmake. A couple of other issues (one is with sqlite, another is with jpeg) have also been addressed.

**Comment by gunan** [comment]
I am not sure about bazel's proper packaging on linux, however we have to support bazel, because internally there is nothing but bazel for us.
Supporting multiple build systems in parallel is also slowing down our development, in not so distant future we will stop actively maintaining the cmake files. Just stating this as our plan as a warning for you.
You are welcome to keep our cmake files working, but just warning you. If you depend on them you will need to maintain our cmake files.

Another warning is, soon we will split most of contrib out of our repository, so cmake files wont even live under core tensorflow repository. But again, you are welcome to maintain them under the new location they will be at.

**Comment by rasitsimsek** [comment]
What ist about another platforms like Solaris or AIX. I think with bazel only build scripts there is no way to build Tensorflow except with CMake. Bazel seems to me Linux centroid.

**Comment by dmacvicar** [comment]
@yongtang Thanks a lot for PR #13145 
I plan to upload my branch with support for system sqlite3, jpeg and others after some cleanups.
It can then be enhanced from there.

One cleanup that may be it makes sense to be done before adding support for system libraries is to rename variables like `jemalloc_STATIC_LIBRARIES` to something like `jemalloc_ext_LIBRARIES` as they will not always be dynamic.

Another is to cleanup targets like `gif_copy_headers_to_destination` so that they are not depended outside of the module, as the system library version of it would not have such thing.

**Comment by johnsrude** [comment]
I think that this support would also need to be extended to Windows.  As referenced at https://github.com/tensorflow/tensorflow/issues/13962 Windows builds of TensorFlow are only supported on CMake, which is what Google uses to build the Windows Python packages.

**Comment by johnsrude** [comment]
@gunan  if you are removing CMake support, when are you going to add bazel builds for TF for Windows?
 https://github.com/tensorflow/tensorflow/issues/13962 states that CMake is what you use to build the Windows Python wheels.  That build takes fives hours and is not incremental.

**Comment by myungjoo** [comment]
Started looking at the external library issue. As a probing point, I've chosen zlib and sent #15382 . I'll continue making other external libraries as "system libraries" if #15382 is satisfiable.

**Comment by rongjiecomputer** [comment]
@johnsrude Tensorflow can already be built with Bazel on Windows, incremental build supported and is much faster than CMake + MSBuild (took me 3 hours in my low-end laptop).

CMake + Ninja will be much faster than CMake + MSBuild as Ninja can do incremental build, probably on par with Bazel. However, for some unknown reason Tensorflow's CMake build scripts are so tuned to MSBuild that Ninja simply does not work.

**Comment by sylvain-bougnoux** [comment]
If you want fast compilation with MS, the best thing to do is precompiled header. It will compile much much faster, as a lot of object files are just a few lines with a huge header.

**Comment by godrxz** [comment]
I'm working on the same thing, any updates? @dmacvicar

**Comment by rongjiecomputer** [comment]
@boegel If it brings any comfort, building Tensorflow with Bazel on Windows is even more painful. Regularly breaks, then few external contributors or someone from Bazel team will come to fix the build. Repeat cycle.

Someone made it possible to build Tensorflow with CMake + Ninja (see https://github.com/tensorflow/tensorflow/pull/16763)!

**Comment by tgamblin** [comment]
I'll just chime in from the Spack side to say that everything @boegel said is true.  He also gave a nice talk -- I was there 😄.  We package software for Livermore Computing (i.e., all of [these machines](https://hpc.llnl.gov/hardware/platforms)), for other large DOE HPC centers, and for HPC sites outside the DOE.  We have a lot of people who want to use TensorFlow, but it's not maintainable, and our attempts to package TF stagnated because of all the work involved.  We basically need to be TF developers to package the thing.

The main issues we have are:
1. Swapping compilers into the build with Bazel is a giant pain.  Spack wants to be able to customize that.  Bazel (last I checked) wants the compiler to be in *exactly* its preferred location, which is hardly ever the case on HPC machines.  There are a lot of things about Bazel that are incompatible with the way things need to be built for OSS projects.
2. We tried the contributed CMake build to address this.  It's more manageable, but the way it used external project at the time didn't provide a way to use an existing installation of the external projects.  So we had to put some nasty hacks in to point the TF build at the spack-built dependencies (see spack/spack#3244).
3. To further complicate things, TF relies not only on the *installed* artifacts from its dependencies, but also internal libraries and headers that aren't usually installed.  It builds its dependencies and grabs stuff out of the build directory, not the install directory.  So even if you have properly packaged versions of the dependencies, you may not have all the things that TF needs. 

Maybe some of this stuff has changed -- If so, I'd love to take a look at this again when I have some time.  I am currently hesitant to do this, since I read here that the TF team was going to ditch the contributed builds, and Bazel is too painful to support in Spack.  See [#6923 (comment)](https://github.com/tensorflow/tensorflow/issues/6923#issuecomment-273671241) where @jart describes what we'd need to do to use Bazel.

For now, we just have a person at Livermore Computing who maintains patches on the Bazel build and updates them (very tediously) when users ask for a newer TF version.

**Comment by ewilderj** [comment]
Hi everyone on this thread: I want to let you know about a new group to address these issues: [tf-distribute](https://groups.google.com/a/tensorflow.org/forum/#!forum/tf-distribute). 

We created it because we hear the pain on these issues and wanted to get the community together so we can come to a solution.

There are [two proposals for a way forward](https://docs.google.com/document/d/1yWIrKZ8-42ZSztYz-TUJQETe_UmSdg0KJzBXMitTQXE/edit) that preserves the ability to build with CMake, please read this doc and comment, and join the discussion on the mailing list.

Basically, we want to address this pain, but we need community involvement too to help this work. I welcome your comments: we certainly don't want to be the ongoing joke in the packaging world, @boegel ;)

**Comment by cdluminate** [comment]
A hacky work in progress https://github.com/tensorflow/tensorflow/pull/21699 . However I'm only able to take care of Debian unstable/experimental.

**Comment by ewilderj** [comment]
So far, SIG Build found nobody interested in keeping CMake alive. Please see this thread https://groups.google.com/a/tensorflow.org/forum/#!topic/build/9vKi3ceP2ZI and others. If you are interested, it's the right time to jump into the discussion.

**Comment by cdluminate** [comment]
@ewilderj Unfortunately I totally lost interest in the cmake build too. I was trying build TF with a python+ninja build system from scratch.

**Comment by ewilderj** [comment]
It's worth taking a look at @perfinion's packaging work for Gentoo. I believe now that Bazel can take care of the use case in this bug. @dmacvicar I'd be interested to know if your requirements can be now met with bazel, and if not what more work needs to happen in SIG Build.

**Comment by cdluminate** [comment]
I checked Gentoo's ebuild for tensorflow. It uses bazel as the build system. However the problem is that bazel doesn't meet many debian's requirement hence is very hard to enter debian.

**Comment by perfinion** [comment]
Yeah once you can get bazel into debian then packaging TF is not too bad, its way better now than even a couple months ago. bazel should be able to build from scratch tho, the first ./compile.sh script just runs javac to get the very basics up and goes from there. iirc it doesnt need the bundled JVM like it had before. Once you get to the TF part feel free to ping me and i'll help how I can. :-)

**Comment by arrowd** [comment]
Just wanted to add that FreeBSD Ports is with you on this boat. It is simply impossible to package TensorFlow for FreeBSD, except for pre-building a package manually and redistributing binaries.

Anything has changed wrt to CMake build since last post?

**Comment by ewilderj** [comment]
Nothing much has changed, but we'd be very excited if you joined the SIG Build community and shared FreeBSD's experience https://groups.google.com/a/tensorflow.org/forum/#!forum/build

**Comment by mohantym** [comment]
Hi @dmacvicar !

I see the [CMake documentations](https://www.tensorflow.org/lite/guide/build_cmake) has been updated to build TFLite-runtime across linux, Windows and other OS architectures.
But not sure about building making distribution friendly w.r.to whole Tensorflow codebase.

Thank you!

**Comment by orowith2os** [comment]
This is still an issue, and one that we (Flathub) have recently encountered.

**Comment by Venkat6871** [comment]
Hi **@dmacvicar**,
Is this still an issue for you? Could you check your code with the recent TF versions and let us know whether the issue persists with recent TF versions.

Thank you!

**Comment by hukl** [comment]
+1

**Comment by sushreebarsa** [comment]
Hi,

There are still a few limitations to make cmake build distribution friendly. TensorFlow includes some dependencies within its source tree. This can cause issues with Linux distribution package managers that prefer system-wide installations. Package managers might not recognize these bundled dependencies, leading to missing libraries during installation.

Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow version with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the release notes to stay up to date with the latest developments which are happening in the Tensorflow space.
